Sources/SwiftProtobuf/ProtobufTextDecoding.swift

```diff
@@ -0,0 +1,1229 @@
+                                    let high = scalar - UInt32(0xD800)
+                                    let low = follower - UInt32(0xDC00)
+                                    let composed = UInt32(0x10000) + high << 10 + low
+                                    if let char = UnicodeScalar(composed) {
+                                        result.append(String(char))
+                                    } else {
+                                        // Composed value is not valid
+                                        return nil
+                                    }
+                                } else {
+                                    // high surrogate was not followed by low
+                                    return nil
+                                }
+                            } else {
+                                // high surrogate not followed by unicode hex escape
+                                return nil
+                            }
+                        }
+                    } else {
+                        // Broken unicode escape
+                        return nil
+                    }
+                default:
+                    // Unrecognized backslash escape
+                    return nil
+                }
+            } else {
+                // Input ends in backslash
+                return nil
+            }
+        default:
+            result.append(c)
+        }
+    }
+    // Unterminated quoted string
+    return nil
+}
+
+// Protobuf JSON allows integers (signed and unsigned) to be
+// coded using floating-point exponential format, but does
+// reject anything that actually has a fractional part.
+// This code converts numbers like 4.294967e9 to a standard
+// integer decimal string format 429496700 using only pure
+// textual operations (adding/removing trailing zeros, removing
+// decimal point character).  The result can be handed
+// to IntMax(string:) or UIntMax(string:) as appropriate.
+//
+// Returns an array of Character (to make it easy for clients to
+// check specific character values) or nil if the provided string
+// cannot be normalized to a valid integer format.
+//
+// Here are some sample inputs and outputs to clarify what this function does:
+//   = "0.1.2" => nil (extra period)
+//   = "0x02" => nil (invalid 'x' character)
+//   = "4.123" => nil (not an integer)
+//   = "012" => nil (leading zero rejected)
+//   = "0" => "0" (bare zero is okay)
+//   = "400e-1" => "40" (adjust decimal point)
+//   = "4.12e2" => "412" (adjust decimal point)
+//   = "1.0000" => "1" (drop extraneous trailing zeros)
+//
+// Note: This does reject sequences that are "obviously" out
+// of the range of a 64-bit integer, but that's just to avoid
+// crazy cases like trying to build million-character string for
+// "1e1000000".  The client code is responsible for real range
+// checking.
+//
+private func normalizeIntString(_ s: String) -> [Character]? {
```

Comment:
In #169, I completely rebuilt the numeric handling.  In particular, this was needed to correctly deal with octal and hexadecimal integer formats:  text format allows them for integer fields, rejects them for floating-point fields.